File: stack/daily_temperatures.739.cpp
#include <vector>
#include <stack>
using std::vector;
using std::stack;

// Monotonic stack
// O(n) time
// O(n) space
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::stack<int> mstack; // store the indices
        int n = temperatures.size();
        std::vector<int> deltas(n, 0);
        for (int i = 0; i < n; ++i) {
            while (!mstack.empty() && temperatures[i] > temperatures[mstack.top()]) {
                // first check !mstack.empty(), then check whether this temperature is higher
                int j = mstack.top();
                mstack.pop();
                deltas[j] = i - j;
            }
            mstack.push(i);
        }
        return deltas;
    }
};

// use std::vector instead of std::stack
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> deltas(n, 0);
        vector<int> stk;
        stk.reserve(n); //
        for (int i = 0; i < n; ++i) {
            while (!stk.empty() && temperatures[i] > temperatures[stk.back()]) {
                int j = stk.back();
                stk.pop_back();
                deltas[j] = i - j;
            }
            stk.push_back(i);
        }
        return deltas;
    }
};

// Brute Force
// O(n^2) time
// O(n) space
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        std::vector<int> deltas(n, 0);
        for (int i = 0; i < n - 1; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (temperatures[j] > temperatures[i]) {
                    deltas[i] = j - i;
                    break;
                }
            }
        }
        return deltas;
    }
};

File: stack/daily_temperatures.739.py
from typing import List

# my solution
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        deltas = [0] * len(temperatures)
        st = []
        for i, temp in enumerate(temperatures):
            while st:
                top_j = st[-1]
                if temp > temperatures[top_j]:
                    deltas[top_j] = i - top_j
                    st.pop()
                else:
                    break
            st.append(i)
        return deltas

# Monotonic stack
# Time: O(n). Each index is pushed and popped at most once (amortized analysis).
# Space: O(n). Auxiliary space for the monotonic stack, which in the worst case (strictly decreasing input) holds all n indices.
# 单调栈问题标准模板？ while stack and 条件: pop
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        deltas = [0] * len(temperatures)
        mstack = []
        for i, temp in enumerate(temperatures):
            while mstack and temp > temperatures[mstack[-1]]:
                top_j = mstack.pop()
                deltas[top_j] = i - top_j
            mstack.append(i)
        return deltas